fix data type logging

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53148 clang format reducer and logger files
* **#53162 fix data type logging**
* #53149 add is_multi_device_module logging field
* #53145 allow users to set sample rate and add per iteration latency breakdowns

it is possible there are multiple data types in mixed precision training, so log data types as a list of data type names.

Differential Revision: [D26769256](https://our.internmc.facebook.com/intern/diff/D26769256/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26769256/)!

zhaojuanmao committed:
fix data type logging

it is possible there are multiple data types in mixed precision training, so log data types as a list of data type names.

Differential Revision: [D26769256](https://our.internmc.facebook.com/intern/diff/D26769256/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26769256/)!

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/logger.cpp:
```diff
@@ -43,9 +43,14 @@ void Logger::set_env_variables() {
   ddp_logging_data_->total_parameter_size_bytes = 0;
-  for (const auto& t : reducer_->replicas_[0]) {
+  std::set<std::string> unique_dtypes;
+  for (auto t : reducer_->replicas_[0]) {
```
Nit: const auto& t (like the code before this change)

zhaojuanmao committed:
Update on "fix data type logging"


it is possible there are multiple data types in mixed precision training, so log data types as a list of data type names.

Differential Revision: [D26769256](https://our.internmc.facebook.com/intern/diff/D26769256/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26769256/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "fix data type logging"


it is possible there are multiple data types in mixed precision training, so log data types as a list of data type names.

Differential Revision: [D26769256](https://our.internmc.facebook.com/intern/diff/D26769256/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26769256/)!

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -65,9 +65,14 @@ void Logger::set_env_variables() {
+  for (auto t : reducer_->replicas_[0]) {
     ddp_logging_data_->total_parameter_size_bytes +=
         t.numel() * t.element_size();
+    unique_dtypes.insert(std::string(t.dtype().name()));
```
Could we remove `unique_dtypes` and just do `ddp_logging_data_->dtypes.push_Back(std::string(t.dtype().name())` directly in this loop?

zhaojuanmao committed:
Update on "fix data type logging"


it is possible there are multiple data types in mixed precision training, so log data types as a list of data type names.

Differential Revision: [D26769256](https://our.internmc.facebook.com/intern/diff/D26769256/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26769256/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "fix data type logging"


it is possible there are multiple data types in mixed precision training, so log data types as a list of data type names.

Differential Revision: [D26769256](https://our.internmc.facebook.com/intern/diff/D26769256/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26769256/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "fix data type logging"


it is possible there are multiple data types in mixed precision training, so log data types as a list of data type names.

Differential Revision: [D26769256](https://our.internmc.facebook.com/intern/diff/D26769256/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26769256/)!

[ghstack-poisoned]
